### PR TITLE
[REF] mrp: state refactoring

### PR DESCRIPTION
**Pull request by amoyaux**
The purpose of this PR is to allow more flexibility on Manufacturing.
It will allows:
- Pull and Push rules are no more directly trigger. It will require a confirmation from the user in order to launch them. (except when it comes from another rule)
- The user will be able to add extra components. Those components will follow the classic flow (trigger rules, will be present in workorders and produce wizard, ...)
- State will be computed and will not require explicit write
- Advanced filter

Task 1889384

### CODE REVIEW

File: addons/mrp/models/mrp_production.py [hunk 1]
```diff
@@ -98,6 +98,34 @@ def _get_default_location_dest_id(self):
...
+             " * To Close: The production is done, the MO has to be closed.\n"
+             " * Done: The MO is closed, the stock moves are posted. \n"
+             " * Cancelled: has been cancelled, can't be confirmed anymore.")
+    reservation_state = fields.Selection([
+        ('confirmed', 'Waiting'),
+        ('assigned', 'Ready'),
+        ('waiting', 'Waiting Another Operation')],
+        string='Material availability',
+        compute='_compute_state', copy=False, index=True, readonly=True,
+        store=True, track_visibility='onchange',
```
Comment (mreficent) [inline addons/mrp/models/mrp_production.py]:
```python
oldname='availability'
```

File: addons/mrp/models/mrp_production.py [hunk 2]
```diff
@@ -517,7 +577,7 @@ def button_plan(self):
             quantity = order.product_uom_id._compute_quantity(order.product_qty, order.bom_id.product_uom_id) / order.bom_id.product_qty
             boms, lines = order.bom_id.explode(order.product_id, quantity, picking_type=order.bom_id.picking_type_id)
             order._generate_workorders(boms)
-        return orders_to_plan.write({'state': 'planned'})
+        return True
```
Comment (Whenrow) [inline addons/mrp/models/mrp_production.py]: Are you sure the state will be recomputed ?

File: addons/mrp/views/mrp_production_views.xml [hunk 3]
```diff
@@ -36,17 +35,16 @@
                 <form string="Manufacturing Orders">
                 <header>
                     <field name="consumed_less_than_planned" invisible="1"/>
-                    <button name="button_mark_done" attrs="{'invisible': ['|', ('check_to_done', '=', False), ('consumed_less_than_planned', '=', True)]}" string="Mark as Done" type="object" class="oe_highlight"/>
-                    <button name="button_mark_done" attrs="{'invisible': ['|', ('check_to_done', '=', False), ('consumed_less_than_planned', '=', False)]}" string="Mark as Done" type="object" class="oe_highlight" confirm="You have consumed less material than what was planned. Are you sure you want to close this MO?"/>
-                    <button name="action_assign" attrs="{'invisible': ['|', ('availability', 'in', ('assigned', 'none')), ('is_locked', '=', False)]}" string="Check availability" type="object" class="oe_highlight"/>
+                    <button name="button_mark_done" attrs="{'invisible': ['|', ('state', '!=', 'to_close'), ('consumed_less_than_planned', '=', True)]}" string="Mark as Done" type="object" class="oe_highlight"/>
```
Comment (Whenrow) [inline addons/mrp/views/mrp_production_views.xml]: what if state = cancel ?

File: addons/mrp/models/mrp_production.py [hunk 4]
```diff
@@ -494,9 +558,49 @@ def _update_raw_move(self, bom_line, line_data):
                 move[0].unlink()
             return move[0], old_qty, quantity
         else:
-            move = self._generate_raw_move(bom_line, line_data)
+            move_values = self._get_move_raw_values(bom_line, line_data)
```
Comment (Whenrow) [inline addons/mrp/models/mrp_production.py]: not _get_moveS_raw_values ?

**Comment by sle-odoo** [comment]
robodoo r+

**Comment by mreficent** [comment]
conflicts

**Comment by sle-odoo** [comment]
robodoo r+

**Comment by sle-odoo** [comment]
robodoo retry

**Pull request closed without merging**